Network Usage event model adjustments

winterhazel opened the pull request:
### Description

This PR fixes the issues with the network-related Usage event model described in #10687:

- `NETWORK.CREATE` is not published when the network is created. Instead, it is published after the network is implemented, which does not match what Usage expects and can result in duplicate `NETWORK.CREATE`s
- `NETWORK.UPDATE` is published to indicate that the network is implemented only when the method that implements the network is called a second time, such as when deploying a second VM on the network, which does not make much sense
- no event is published when the network shuts down and goes from `Implemented` to `Allocated`

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI
- [ ] test (unit or integration test code)

### Feature/Enhancement Scale or Bug Severity

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [X] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to -->

#### How did you try to break this feature and the system with this change?


Isolated network:

1. I created a network: a `NETWORK.CREATE` was published
2. I deployed a VM on the network: a `NETWORK.UPDATE` (`Implemented`) was published
3. I deployed a second VM on the network: no network-related event was published
4. I shut down both VMs and waited until the network was shut down by the garbage collector: a `NETWORK.UPDATE` (`Allocated`) was published
5. I started one of the VMs: a `NETWORK.UPDATE` (`Implemented`) was published
6. I restarted the network (with and without clean-up): no network-related events were published
7. I destroyed the network: a `NETWORK.UPDATE` (`Allocated`) and a `NETWORK.DELETE` were published

Error on isolated network deployment:

8. I changed `router.template.kvm` to a template that does not exist so that the network implementation would fail, created a new network, and tried to deploy a VM on it: only `NETWORK.CREATE` was published because the implementation failed

Persistent networks:

9. I created a persistent network: a `NETWORK.CREATE` and a `NETWORK.UPDATE` (`Implemented`) were published
10. I deployed another VM on the network: no network-related events were published
11. I destroyed the network: a `NETWORK.UPDATE` (`Allocated`) and a `NETWORK.DELETE` were published

Shared networks:

12. I deployed a shared network and a VM on it: only the `NETWORK.CREATE` was published (which is expected, as there was no state change)
13. I destroyed the network: a `NETWORK.DELETE` was published

L2 networks:

14. I deployed a L2 network: a `NETWORK.CREATE` was published
15. I deployed a VM on the network: a `NETWORK.UPDATE` (`Implemented`) was published
16. I destroyed the network: a `NETWORK.UPDATE` (`Allocated`) and a `NETWORK.DELETE` were published

Kubernetes:

17. I deployed a Kubernetes cluster without specifying the network: a `NETWORK.CREATE` and a `NETWORK.UPDATE` (`Implemented`) were published

winterhazel commented:
@blueorangutan package

blueorangutan commented:
@winterhazel a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✖️ el8 ✖️ el9 ✖️ debian ✖️ suse15. SL-JID 13123

sureshanaparti commented:
@blueorangutan package

blueorangutan commented:
@sureshanaparti a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 13141

DaanHoogland commented on engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:
```diff
@@ -4780,6 +4781,27 @@ public void unmanageNics(VirtualMachineProfile vm) {
         }
     }
 
+    @Override
+    public void publishNetworkCreation(Network network) {
+        UsageEventUtils.publishUsageEvent(EventTypes.EVENT_NETWORK_CREATE, network.getAccountId(), network.getDataCenterId(),
+                network.getId(), network.getName(), network.getNetworkOfferingId(), null, null, null, network.getState().name(),
+                network.getUuid());
+    }
+
+    @Override
+    public void publishNetworkUpdate(Network network) {
+        UsageEventUtils.publishUsageEvent(EventTypes.EVENT_NETWORK_UPDATE, network.getAccountId(), network.getDataCenterId(),
+                network.getId(), network.getName(), network.getNetworkOfferingId(), null, network.getState().name(),
+                Network.class.getName(), network.getUuid(), true);
+    }
+
+    @Override
+    public void publishNetworkDeletion(Network network) {
+        UsageEventUtils.publishUsageEvent(EventTypes.EVENT_NETWORK_DELETE, network.getAccountId(), network.getDataCenterId(),
+                network.getId(), network.getName(), network.getNetworkOfferingId(), null, null, null,
+                Network.class.getName(), network.getUuid());
+    }
+
```
can these be default implementations in the `interface NetworkOrchestrationService`? Maybe we do not want the `NetworkEventUtils` in there, but only a static method is called on that anyway.

DaanHoogland reviewed (approved):
clgtm, I think we can think of some follow up as discussed, but this certainly addresses the initial issue.

DaanHoogland commented:
@rajujith , can you see if this alleviates any of your usage issues?

DaanHoogland commented:
> @DaanHoogland This solves the problem for time being. But for the long term we may want to record different usage events like you mentioned here [#10687 (comment)](https://github.com/apache/cloudstack/issues/10687#issuecomment-2804174489) and create two separate usage types like we already have for running and allocated VM.

thanks @rajujith , can you test this with your faulty use-case scenario?

@winterhazel would you agree with that requirement?

winterhazel commented:
> @DaanHoogland This solves the problem for time being. But for the long term we may want to record different usage events like you mentioned here [#10687 (comment)](https://github.com/apache/cloudstack/issues/10687#issuecomment-2804174489) and create two separate usage types like we already have for running and allocated VM.

@rajujith after looking into and understanding how the network Usage event model works (or was supposed to work), I do not think anymore that it is necessary to add separate events for `IMPLEMENT` and `STOP`, at least in the context of Usage. The network's new state is in included in the `NETWORK.UPDATE` event. This can be used to identify when a network transitioned from allocated to implemented and vice-versa.

Also, we include the network's state during that period in the usage record. This can be used to distinguish between allocated and implemented usage.

When the running and allocated virtual machine usage types were introduced, we did not have the state column. I'm not sure why it was introduced that way (two different usage types instead of including the state in the usage record), so genuine question: is there an advantage of having two separate usage types just to distinguish between allocated and implemented? For me, it seems unnecessary.

winterhazel commented:
@blueorangutan package

blueorangutan commented:
@winterhazel a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

rajujith commented:
> > @DaanHoogland This solves the problem for time being. But for the long term we may want to record different usage events like you mentioned here [#10687 (comment)](https://github.com/apache/cloudstack/issues/10687#issuecomment-2804174489) and create two separate usage types like we already have for running and allocated VM.
> 
> @rajujith after looking into and understanding how the network Usage event model works (or was supposed to work), I do not think anymore that it is necessary to add separate events for `IMPLEMENT` and `STOP` (at least in the context of Usage. Maybe it makes sense for the standard events). The network's new state is in included in the `NETWORK.UPDATE` event. This can be used to identify when a network transitioned from allocated to implemented and vice-versa.
> 
> Also, we include the network's state in the usage record during that period. This can be used to distinguish between allocated and implemented usage.
> 
> When the running and allocated virtual machine usage types were introduced, we did not have the state column. I'm not sure why it was introduced that way (two different usage types instead of including the state in the usage record), so genuine question: is there an advantage of having two separate usage types just to distinguish between allocated and implemented? For me, it seems unnecessary.

@winterhazel  I can only imagine a scenario where the operator would want to bill  allocated and Implemented networks at a different rate or as separate products. If these usages can be differentiated in the API response that should be enough. However separate usage type makes a clear distinction in my head, may be its not necessary.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 13171

rohityadavcloud commented:
@blueorangutan test

blueorangutan commented:
@rohityadavcloud a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian Build Failed (tid-13148)<b/>

blueorangutan commented:
<b>[SF] Trillian Build Failed (tid-13150)<b/>

rajujith reviewed (approved):
LGTM.

Based on my tests the duplicate events are no longer created. No more duplicate usage records are being generated as well.

```
mysql> select * from usage_event where  resource_id=204 and type in ('NETWORK.CREATE','NETWORK.DELETE','NETWORK.UPDATE');
+----+----------------+------------+---------------------+---------+-------------+---------------+-------------+-------------+------+---------------+-----------+--------------+
| id | type           | account_id | created             | zone_id | resource_id | resource_name | offering_id | template_id | size | resource_type | processed | virtual_size |
+----+----------------+------------+---------------------+---------+-------------+---------------+-------------+-------------+------+---------------+-----------+--------------+
|  2 | NETWORK.CREATE |          2 | 2025-04-25 08:03:40 |       1 |         204 | net1          |          10 |        NULL | NULL | NULL          |         1 |         NULL |
|  6 | NETWORK.UPDATE |          2 | 2025-04-25 08:04:57 |       1 |         204 | net1          |          10 |        NULL | NULL | Implemented   |         1 |         NULL |
| 11 | NETWORK.UPDATE |          2 | 2025-04-25 08:08:58 |       1 |         204 | net1          |          10 |        NULL | NULL | Allocated     |         1 |         NULL |
| 12 | NETWORK.UPDATE |          2 | 2025-04-25 08:11:14 |       1 |         204 | net1          |          10 |        NULL | NULL | Implemented   |         1 |         NULL |
| 19 | NETWORK.UPDATE |          2 | 2025-04-25 08:14:31 |       1 |         204 | net1          |          10 |        NULL | NULL | Allocated     |         1 |         NULL |
| 21 | NETWORK.DELETE |          2 | 2025-04-25 08:14:33 |       1 |         204 | net1          |          10 |        NULL | NULL | NULL          |         1 |         NULL |
+----+----------------+------------+---------------------+---------+-------------+---------------+-------------+-------------+------+---------------+-----------+--------------+
6 rows in set (0.00 sec)

mysql> select account_id,usage_id,usage_type,start_date,end_date from cloud_usage.cloud_usage where usage_id=204;
+------------+----------+------------+---------------------+---------------------+
| account_id | usage_id | usage_type | start_date          | end_date            |
+------------+----------+------------+---------------------+---------------------+
|          2 |      204 |         30 | 2025-04-25 07:30:00 | 2025-04-25 08:29:59 |
+------------+----------+------------+---------------------+---------------------+
1 row in set (0.00 sec)

```

Pearl1594 commented:
@blueorangutan test

blueorangutan commented:
@Pearl1594 a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-13143)</b>
Environment: kvm-ol8 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 47513 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr10755-t13143-kvm-ol8.zip
Smoke tests completed. 132 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_02_list_cpvm_vm | `Failure` | 0.06 | test_ssvm.py
test_04_cpvm_internals | `Failure` | 0.06 | test_ssvm.py

blueorangutan commented:
<b>[SF] Trillian test result (tid-13157)</b>
Environment: kvm-ol8 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 53512 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr10755-t13157-kvm-ol8.zip
Smoke tests completed. 132 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_secure_vm_migration | `Error` | 136.79 | test_vm_life_cycle.py
test_01_secure_vm_migration | `Error` | 136.80 | test_vm_life_cycle.py
